File: toolkit/livedata/src/main/java/br/com/arch/toolkit/livedata/computable/ComputableLiveData.java
package br.com.arch.toolkit.livedata.computable;

import android.arch.lifecycle.LiveData;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import java.lang.ref.WeakReference;
import java.util.concurrent.atomic.AtomicBoolean;

import static br.com.arch.toolkit.livedata.ExecutorUtil.async;

/**
 * Implementation of {@link LiveData} to make it possible compute something on the WorkerThread, and abort or invalidate if needed
 * <p>
 * Will compute the value only when it has at least one observer, in other words, if the liveData instance is active
 *
 * @param <T> The type of the value do compute
 */
public abstract class ComputableLiveData<T> extends LiveData<T> {

    private final Object lock = new Object();
    private final AtomicBoolean computing = new AtomicBoolean(false);
    private final AtomicBoolean computed = new AtomicBoolean(false);

    @Nullable
    private WeakReference<Thread> lastThread = null;

    /**
     * Method executed on the WorkerThread and responsible for compute some data
     * <p>
     * Please, use the {@link LiveData#postValue(Object)} method inside the implementation of this method to post new values into the ComputableLiveData
     */
    @WorkerThread
    abstract void compute();

    /**
     * Method used to abort some ongoing async operation.
     * <p>
     * Example:
     * - You can call the {@link Thread#interrupt()} method if you decided, for some reason, start a new thread in your application
     */
    abstract void abort();

    /**
     * @return true if it is computing some value, false otherwise
     */
    public boolean isRunning() {
        return computing.get();
    }

    /**
     * @return true if it has computed some value, false otherwise
     */
    public boolean hasComputed() {
        return computed.get();
    }

    /**
     * @see ComputableLiveData#isRunning()
     * @see ComputableLiveData#hasComputed()
     */
    public boolean isRunningOrHasComputed() {
        return computed.get() || computing.get();
    }

    /**
     * Invalidate the actual data and compute again
     */
    public void invalidate() {
        if (!hasObservers() && computed.get()) computed.set(false);
        else if (!computing.get()) {
            computed.set(false);
            lastThread = new WeakReference<>(async(this::executeRunnable));
        }
    }

    /**
     * Interrupt the ongoing computing Thread and call {@link ComputableLiveData#abort()}
     */
    public void interrupt() {
        if (lastThread != null && lastThread.get() != null) {
            lastThread.get().interrupt();
        }
        abort();
        computing.set(false);
        computed.set(false);
    }

    @Override
    protected void onActive() {
        super.onActive();
        if (!computed.get()) invalidate();
    }

    private void executeRunnable() {
        synchronized (lock) {
            if (computed.get()) return;

            try {
                computing.set(true);
                compute();
                computed.set(true);
            } catch (final Exception error) {
                computed.set(false);
            } finally {
                computing.set(false);
            }
        }
    }
}

File: toolkit/multi/event-observer/src/androidMain/kotlin/br/com/arch/toolkit/livedata/computable/ComputableLiveData.kt
package br.com.arch.toolkit.livedata.computable

import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Implementation of [LiveData] to make it possible compute something on the WorkerThread, and abort or invalidate if needed
 *
 *
 * Will compute the value only when it has at least one observer, in other words, if the liveData instance is active
 *
 * @param <T> The type of the value do compute
</T> */
abstract class ComputableLiveData<T> : LiveData<T>() {

    private val lock = Any()
    private val computing = AtomicBoolean(false)
    private val computed = AtomicBoolean(false)
    private var lastThread: WeakReference<Thread?>? = null

    /**
     * Method executed on the WorkerThread and responsible for compute some data
     *
     *
     * Please, use the [LiveData.postValue] method inside the implementation of this method to post new values into the ComputableLiveData
     */
    @WorkerThread
    abstract fun compute()

    /**
     * Method used to abort some ongoing async operation.
     *
     *
     * Example:
     * - You can call the [Thread.interrupt] method if you decided, for some reason, start a new thread in your application
     */
    abstract fun abort()

    /**
     * @return true if it is computing some value, false otherwise
     */
    val isRunning: Boolean
        get() = computing.get()

    /**
     * @return true if it has computed some value, false otherwise
     */
    fun hasComputed() = computed.get()

    /**
     * @see ComputableLiveData.isRunning
     * @see ComputableLiveData.hasComputed
     */
    val isRunningOrHasComputed: Boolean
        get() = computed.get() || computing.get()

    /**
     * Invalidate the actual data and compute again
     */
    fun invalidate() {
        if (!hasObservers() && computed.get()) {
            computed.set(false)
        } else if (!computing.get()) {
            computed.set(false)
            lastThread = WeakReference(runOnNewThread { executeRunnable() })
        }
    }

    /**
     * Interrupt the ongoing computing Thread and call [ComputableLiveData.abort]
     */
    fun interrupt() {
        if (lastThread != null && lastThread!!.get() != null) {
            lastThread!!.get()!!.interrupt()
        }
        abort()
        computing.set(false)
        computed.set(false)
    }

    override fun onActive() {
        super.onActive()
        if (!computed.get()) invalidate()
    }

    private fun executeRunnable() {
        synchronized(lock) {
            if (computed.get()) return
            runCatching {
                computing.set(true)
                compute()
                computed.set(true)
            }.onFailure {
                computed.set(false)
            }
            computing.set(false)
        }
    }

    /**
     * Execute a Async Thread block
     *
     * @param block The block containing the instructions to make on the WorkerThreat
     * @return The executing Thread instance
     */
    private fun runOnNewThread(block: Runnable): Thread {
        val thread = Thread(block)
        thread.start()
        return thread
    }
}
